internal/ctxerror/ctxerror.go

```diff
@@ -0,0 +1,124 @@
+// Package ctxerror provides a context-aware error facility.
+//
+// Inspired by log15-style (semi-)structured logging,
+// it also provides a log15 bridge.
+package ctxerror
+
+//go:generate mockgen -source ctxerror.go -destination mock/ctxerror.go
+
+import (
+	"fmt"
+)
+
+// CtxError is a context-aware error container.
+type CtxError interface {
+	// Error returns a fully formatted message, with context info.
+	Error() string
+
+	// Message returns the bare error message, without context info.
+	Message() string
+
+	// Contexts returns message contexts.
+	// Caller shall not modify the returned map.
+	Contexts() map[string]interface{}
+
+	// WithCause chains an error after the receiver.
+	// It returns the merged/chained instance,
+	// where the message is "<receiver.Message>: <c.Message>",
+	// and with contexts merged (ones in c takes precedence).
+	WithCause(c error) CtxError
+}
+
+type ctxError struct {
+	msg string
+	ctx map[string]interface{}
+}
+
+// New creates and returns a new context-aware error.
+func New(msg string, ctx ...interface{}) CtxError {
+	e := &ctxError{msg: msg, ctx: make(map[string]interface{})}
+	e.updateCtx(ctx...)
+	return e
+}
+
+func (e *ctxError) updateCtx(ctx ...interface{}) {
+	var name string
+	if len(ctx)%2 == 1 {
+		ctx = append(ctx, nil)
+	}
+	for idx, value := range ctx {
+		if idx%2 == 0 {
+			name = value.(string)
+		} else {
+			e.ctx[name] = value
+		}
+	}
+}
+
+// Error returns a fully formatted message, with context info.
+func (e *ctxError) Error() string {
+	s := e.msg
+	for k, v := range e.ctx {
+		s += fmt.Sprintf(", %s=%#v", k, v)
+	}
+	return s
+}
+
+// Message returns the bare error message, without context info.
+func (e *ctxError) Message() string {
+	return e.msg
+}
+
+// Contexts returns message contexts.
+// Caller shall not modify the returned map.
+func (e *ctxError) Contexts() map[string]interface{} {
+	return e.ctx
+}
+
+// WithCause chains an error after the receiver.
+// It returns the merged/chained instance,
+// where the message is “<receiver.Message>: <c.Message>”,
+// and with contexts merged (ones in c takes precedence).
+func (e *ctxError) WithCause(c error) CtxError {
+	r := &ctxError{msg: e.msg + ": ", ctx: make(map[string]interface{})}
+	for k, v := range e.ctx {
+		r.ctx[k] = v
+	}
+	switch c := c.(type) {
```

Comment:
lgtm overall, i'm just curious c is already an error - why is it here?

Reply:
@mikedoan, that's a [variable-shadowing type switch idiom](https://golang.org/doc/effective_go.html#type_switch).  The left-hand-side variable `c` is local to the switch block scope, and shadows the original `c` on the right-hand side, which is in the enclosing scope.  Since it is a type switch, the LHS gets the same value as the RHS, so this has the effect of dynamically typecasting `c` down to more specific types:

```go
var v interface{}
v = /* something */
switch v := v.(type) {
// In all of the cases below,
// the shadowing (LHS) v has the same value as the original (RHS) v.
case TypeA:
    // v is downcasted to TypeA in this case
    /* ... */
case TypeB:
    // v is downcasted to TypeB in this case
    /* ... */
default:
    // v is still interface{} in this case
    /* ... */
}
```